### PR TITLE
Corrected MCMC sampling in BPMF algorithm

### PR DESCRIPTION
**Pull request by LoryPack**
The previous code did not implement correctly MCMC sampling in the BPMF algorithm. In fact, at every step it computed the predictions basing on the current value of the feature matrices, and used it to estimate the RMSE. This has no meaning from the MCMC point of view, whose purpose is to sample the feature matrices from the correct distributions in order to estimate the integral through which the rating distribution is computed. 
Instead, the correct approach (see Eq. 10 in "Bayesian Probabilistic Matrix Factorization using MCMC") entails averaging the predictions at every time step to get a final prediction and compute the RMSE. Essentially, the predicted value itself does not depend only on the last extracted value for the feature matrices, but on the whole chain. Having modified this, the RMSE for both the train and test set with BPMF improves (you can see it in the attached image
[1M-wrong_correct_sampling.pdf](https://github.com/chyikwei/recommend/files/1998644/1M-wrong_correct_sampling.pdf)
)

### CODE REVIEW

**Comment by chyikwei** [comment]
Thanks for the PR! 
You are right. we need to average the prediction values in each sampling steps. 
For the code, I would like to clean it up before merge though.

File: recommend/bpmf.py [hunk 1]
```diff
@@ -60,57 +60,73 @@ def __init__(self, n_user, n_item, n_feature, beta=2.0, beta_user=2.0,
...
         self.item_features_ = 0.3 * self.rand_state.rand(n_item, n_feature)
 
         # data state
         self.mean_rating_ = None
         self.ratings_csr_ = None
         self.ratings_csc_ = None
 
-    def fit(self, ratings, n_iters=50):
-        """training models"""
+    def fit(self, ratings, test_ratings=None, n_iters=50):
```
Comment (chyikwei) [inline recommend/bpmf.py]: pass `test_ratings` in the training step doesn't make much sense in the api point of view.
I think it is better to keep the average user/item matrix and use it in prediction directly?

File: recommend/bpmf.py [hunk 2]
```diff
@@ -60,58 +60,148 @@ def __init__(self, n_user, n_item, n_feature, beta=2.0, beta_user=2.0,
...
-
+        self.ratings_csc_ = self.ratings_csr_.tocsc() 
+    
+        # lists in where I will save the rmse during training
+        train_rmse_list = []
+        if test_ratings is not None: 
+            test_rmse_list = []
+            
+        # define the lists that will contain the total predictions for training and test set
+        train_preds_total = np.zeros(len(ratings)).tolist()
```
Comment (chyikwei) [inline recommend/bpmf.py]: Use numpy array will save more space here.

**Comment by chyikwei** [comment]
yeah. It would be good to add "burn-in" parameter. I am traveling now. will check the code this weekend.

**Comment by chyikweiyau** [comment]
For now I tried to stick to the original API so I added 2 additional variables (`avg_user_features_` and `avg_item_features_`) to track the average features during MCMC steps and use it for prediction.

The update code is here:
https://github.com/chyikwei/recommend/tree/pr_24

I cannot add my commits to your branch directly, so maybe you can merge the change and update this PR. (Then I can merge and close this PR.)

Notes:
- I haven't add the `burn-in` parameter yet. I think it will be better to change that in a new PR.
- Still thinking how to change the API to track test RMSE during MCMC steps. (maybe add a `partial_fit` method)

**Comment by chyikwei** [comment]
For now I tried to stick to the original API so I added 2 additional variables (`avg_user_features_` and `avg_item_features_`) to track the average features during MCMC steps and use it for prediction.

The update code is here:
https://github.com/chyikwei/recommend/tree/pr_24

I cannot add my commits to your branch directly, so maybe you can merge the change and update this PR. (Then I can merge and close this PR.)

Notes:
- I haven't add the `burn-in` parameter yet. I think it will be better to change that in a new PR.
- Still thinking how to change the API to track test RMSE during MCMC steps. (maybe add a `partial_fit` method)

**Comment by chyikwei** [comment]
I create a PR #25 with updated code. close this one now.

**Comment by LoryPack** [comment]
Great, I am sorry I did not have time to do it myself.
